### DISCUSSION TITLE
GMP not working with ASM istioside enabled pod

### DISCUSSION THREAD

**Discussion by sreejesh-radhakrishnan-db** [GoogleCloudPlatform/prometheus-engine#966] [General]
@pintohutch hope you are still the SME on this area. We are having a similar issue where ISTIO is running on 15090 and scrape away all the metric. but my app exposing on 8084 is not getting through to GPM.
I followed your step above and I could see this error :

Get "http://240.13.206.101:8084/actuator/prometheus": read tcp xxx.xxx.xxx.xxx:37516->xxx.xxx.xx.xxx:8084: read: connection reset by peer

basically collector not able to connect to the Application POD.

any help please?

also observed this error on pod -istio sidecar container

 "textPayload": "[2024-05-10T15:48:49.862Z] \"- - -\" 0 NR filter_chain_not_found - \"-\" 0 0 0 - \"-\" \"-\" \"-\" \"-\" \"-\" - - 240.13.xxx.xxx:8084 240.13.xxx.xxx:51396 - -", 

similar to #340 but no solution there

**Comment by pintohutch** [comment]
Hi @sreejesh-radhakrishnan-db - how do you have [TLS configured](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings) for Istio?

**Reply by sreejesh-radhakrishnan-db** [reply]
its STRICT 


---
# PeerAuthentication defines how traffic will be tunneled (or not) to the sidecar.
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default
  namespace: istio-system
spec:
  mtls:
    mode: STRICT # Connection is an mTLS tunnel (TLS with client cert must be presented).

**Reply by sreejesh-radhakrishnan-db** [reply]
@pintohutch   thanks for the reply :-)

**Reply by sreejesh-radhakrishnan-db** [reply]
just to add, what is weird and wonderful is  :-) 

CONTROLLER to ISTIO SIDECAR matrix is working and we are getting all the ISTIO metric collected by GPM.

its just the application specific connection getting terminated or not working from controller

**Reply by pintohutch** [reply]
Hey @sreejesh-radhakrishnan-db,

My personal experience with Istio is limited, however from reading their [docs](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings):
> The control plane, gateway, and Envoy sidecar metrics will all be scraped over cleartext. However, the application metrics will follow whatever [Istio authentication policy](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/) has been configured for the workload.

> If you use STRICT mode, then Prometheus will need to be configured to scrape using Istio certificates as described below.

It then gives instructions on injecting a sidecar container to your Prometheus server.

If you're using [self-deployed collection](https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-unmanaged), this should work. However, if you're using [managed collection](https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-managed), we would need to take a different approach. Which method are you using?

**Reply by sreejesh-radhakrishnan-db** [reply]
thanks @pintohutch    we are using Managed collection. on GKE enterprise 1.27.11-gke.1062001

**Reply by pintohutch** [reply]
@TheSpiritXIII do we have any reference or example for using Istio certs for the new secrets management feature?

**Reply by TheSpiritXIII** [reply]
Hi @sreejesh-radhakrishnan-db I've yet to test it out but GMP needs access to the Istio certificates. I'm not sure if you can get extract them easily, but you can give Istio your own certificates: https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/

By then you should have your Istio certificates in a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/). Finally, [set the fields in the scrape endpoint](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/doc/api.md#monitoring.googleapis.com/v1.TLS).

We are working on public documentation for configuring TLS certificates in general!

**Reply by maxromanovsky** [reply]
@TheSpiritXIII I am trying to use `tls` config, but getting an error `Error from server (BadRequest): error when creating "/xxx/pm.yaml": PodMonitoring in version "v1" cannot be handled as a PodMonitoring: strict decoding error: unknown field "spec.endpoints[0].tls"` with the following manifest.

What am I doing wrong?

```yaml
apiVersion: monitoring.googleapis.com/v1
kind: PodMonitoring
metadata:
  name: test
  namespace: test
spec:
  endpoints:
  - interval: 30s
    path: /
    port: metrics
    tls:
      ca:
        secret:
          namespace: xxx
          name: yyy
          key: zzz.pem
  selector:
    matchLabels:
      app: aaa
```

I see these annotations for `collector` pods:
```
components.gke.io/component-name: managed-prometheus
components.gke.io/component-version: 0.8.3-gke.0
```

Does it need to be upgraded to `v0.10.0`?
I can't find any docs on how it can be upgraded manually without upgrading GKE to 1.27+.

**Reply by TheSpiritXIII** [reply]
@maxromanovsky hello! You need GMP v0.12.0, which is currently rolling out to new and upgraded GKE 1.28 clusters. What version are you currently on?

We plan to eventually backport v0.12.0 to GKE 1.27. GKE 1.26 will be end-of-life next month, so we will likely miss the cut-off.

**Reply by maxromanovsky** [reply]
@TheSpiritXIII we're planning to migrate to GKE 1.27 soon. Any ETA on 1.27 support?
Thanks in advance!

**Reply by TheSpiritXIII** [reply]
There's no ETA yet. By policy, we have to let it "soak" on 1.28 before we can put it on 1.27. I can circle back with the exact 1.27 version once we've backported!
